client/relay-chain-rpc-interface/src/reconnecting_ws_client.rs

```diff
@@ -0,0 +1,390 @@
+	/// - Distribute incoming import, best head and finalization notifications to registered listeners.
+	///   If an error occurs during sending, the receiver has been closed and we remove the sender from the list.
+	/// - Find a new valid RPC server to connect to in case the websocket connection is terminated.
+	///   If the worker is not able to connec to an RPC server from the list, the worker shuts down.
+	async fn run(mut self) {
+		let mut pending_requests = FuturesUnordered::new();
+
+		let Ok(mut client_manager) = ClientManager::new(self.ws_urls.clone()).await else {
+			tracing::error!(target: LOG_TARGET, "No valid RPC url found. Stopping RPC worker.");
+			return;
+		};
+		let Ok(mut subscriptions) = client_manager.get_subscriptions().await else {
+			return;
+		};
+
+		let mut should_reconnect = false;
+		loop {
+			// This branch is taken if the websocket connection to the current RPC server is closed.
+			if should_reconnect {
+				// At this point, all pending requests will return an error since the
+				// websocket connection is dead. So draining the pending requests should be fast.
+				while !pending_requests.is_empty() {
+					if let Some(Err(req)) = pending_requests.next().await {
+						// Put the failed requests into the queue again, they will be processed
+						// once we have a new rpc server to connect to.
+						self.self_sender
+							.send(req)
+							.await
+							.expect("This worker owns the channel; qed");
+					}
+				}
+
+				if let Err(_) = client_manager.connect_to_new_rpc_server().await {
+					tracing::error!(
+						target: LOG_TARGET,
+						"Unable to find valid external RPC server, shutting down."
+					);
+					return
+				};
+
+				let Ok(new_subscriptions) = client_manager.get_subscriptions().await else {
+					tracing::error!(
+						target: LOG_TARGET,
+						"Not able to create streams from newly connected RPC server, shutting down."
+					);
+					return;
+				};
+
+				subscriptions = new_subscriptions;
+				should_reconnect = false;
+			};
+
+			tokio::select! {
+				evt = self.client_receiver.recv() => match evt {
+					Some(RpcDispatcherMessage::RegisterBestHeadListener(tx)) => {
+						self.best_header_listeners.push(tx);
+					},
+					Some(RpcDispatcherMessage::RegisterImportListener(tx)) => {
+						self.imported_header_listeners.push(tx)
+					},
+					Some(RpcDispatcherMessage::RegisterFinalizationListener(tx)) => {
+						self.finalized_header_listeners.push(tx)
+					},
+					Some(RpcDispatcherMessage::Request(method, params, response_sender)) => {
+						pending_requests.push(client_manager.create_request(method, params, response_sender));
+					},
+					None => {
+						tracing::error!(target: LOG_TARGET, "RPC client receiver closed. Stopping RPC Worker.");
+						return;
+					}
+				},
+				should_retry = pending_requests.next(), if !pending_requests.is_empty() => {
```

Comment:
If `pending_requests` is empty this will result in a hot loop. Aka it will always return and thus we will loop all the time until there is some request added.

Reply:
After reading the docs as hinted by @skunert, this if condition is a precondition, aka we will not poll the future when this returns false.